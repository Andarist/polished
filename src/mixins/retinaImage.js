// @flow
import hiDPI from './hiDPI'
import validateModule from '../validation/_validateModule'

/**
 * A helper to generate a retina background image and non-retina
 * background image. The retina background image will output to a HiDPI media query. The mixin uses
 * a _2x.png fileName suffix by default.
 *
 * @example
 * // Styles as object usage
 * const styles = {
 *  ...retinaImage('my-img')
 * }
 *
 * // styled-components usage
 * const div = styled.div`
 *   ${retinaImage('my-img')}
 * `
 *
 * // CSS as JS Output
 * div {
 *   backgroundImage: 'url(my-img.png)',
 *   '@media only screen and (-webkit-min-device-pixel-ratio: 1.3),
 *    only screen and (min--moz-device-pixel-ratio: 1.3),
 *    only screen and (-o-min-device-pixel-ratio: 1.3/1),
 *    only screen and (min-resolution: 144dpi),
 *    only screen and (min-resolution: 1.5dppx)': {
 *     backgroundImage: 'url(my-img_2x.png)',
 *   }
 * }
 */
function retinaImage(
  fileName: string,
  backgroundSize?: string,
  extension?: string = 'png',
  retinaFileName?: string,
  retinaSuffix?: string = '_2x',
) {
  /* istanbul ignore next */
  if (process.env.NODE_ENV !== 'production') {
    if (
      !validateModule('mixins/placeholder.js', {
        // eslint-disable-next-line prefer-rest-params
        arrityCheck: { args: arguments, min: 1, max: 5 },
        typeCheck: [
          {
            param: fileName,
            type: 'string',
            required: 'requires a fileName as its first parameter. However, you did not provide one.',
          },
          { param: backgroundSize, type: 'string' },
          { param: extension, type: 'string' },
          { param: retinaFileName, type: 'string' },
          { param: retinaSuffix, type: 'string' },
        ],
      })
    ) {
      return {}
    }
  }

  // Replace the dot at the beginning of the passed extension if one exists
  const ext = extension.replace(/^\./, '')
  const rFileName = retinaFileName
    ? `${retinaFileName}.${ext}`
    : `${fileName}${retinaSuffix}.${ext}`

  return {
    backgroundImage: `url(${fileName}.${ext})`,
    [hiDPI()]: {
      backgroundImage: `url(${rFileName})`,
      backgroundSize,
    },
  }
}

export default retinaImage
