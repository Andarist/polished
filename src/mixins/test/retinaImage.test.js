// @flow
import retinaImage from '../retinaImage'

describe('retinaImage', () => {
  beforeAll(() => {
    global.console = {
      error: jest.fn(),
      warn: jest.fn(),
      log: global.console.log,
    }
  })

  afterEach(() => {
    // eslint-disable-next-line no-console
    console.error.mockClear()
    // eslint-disable-next-line no-console
    console.warn.mockClear()
  })

  it('should use _2x and png as the default suffix and extension, respectively', () => {
    expect({ ...retinaImage('img') }).toMatchSnapshot()
  })

  it('should set the background-size if one is passed in', () => {
    expect({ ...retinaImage('img', 'cover') }).toMatchSnapshot()
  })

  it('should set the extension if one is passed in', () => {
    expect({ ...retinaImage('img', undefined, 'jpg') }).toMatchSnapshot()
  })

  it('should allow passing in an extension with a dot', () => {
    expect({ ...retinaImage('img', undefined, '.jpg') }).toMatchSnapshot()
  })

  it('should allow passing in a separate filename for the retina version', () => {
    expect({
      ...retinaImage('img', undefined, undefined, 'retina_img'),
    }).toMatchSnapshot()
  })

  it('should allow passing in a separate suffix for the retina version', () => {
    expect({
      ...retinaImage('img', undefined, undefined, undefined, '_5x'),
    }).toMatchSnapshot()
  })

  it('should throw an error when no parameters are passed', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    retinaImage()
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when a non-string value is passed for fileName', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    retinaImage(1)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when a non-string value is passed for backgroundSize', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    retinaImage('img', 100)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })

  it('should throw an error when a non-string value is passed for extension', () => {
    // $FlowIgnoreNextLine since the coming is invalid code, flow complains
    retinaImage('img', undefined, true)
    // eslint-disable-next-line no-console
    expect(console.error.mock.calls).toMatchSnapshot()
  })
})
