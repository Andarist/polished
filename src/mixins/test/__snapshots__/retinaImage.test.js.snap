exports[`retinaImage should allow passing in a separate filename for the retina version 1`] = `
Object {
  "
    @media only screen and (-webkit-min-device-pixel-ratio: 1.3),
    only screen and (min--moz-device-pixel-ratio: 1.3),
    only screen and (-o-min-device-pixel-ratio: 1.3/1),
    only screen and (min-resolution: 125dpi),
    only screen and (min-resolution: 1.3dppx)
  ": Object {
    "backgroundImage": "url(retina_img.png)",
    "backgroundSize": undefined,
  },
  "backgroundImage": "url(img.png)",
}
`;

exports[`retinaImage should allow passing in a separate suffix for the retina version 1`] = `
Object {
  "
    @media only screen and (-webkit-min-device-pixel-ratio: 1.3),
    only screen and (min--moz-device-pixel-ratio: 1.3),
    only screen and (-o-min-device-pixel-ratio: 1.3/1),
    only screen and (min-resolution: 125dpi),
    only screen and (min-resolution: 1.3dppx)
  ": Object {
    "backgroundImage": "url(img_5x.png)",
    "backgroundSize": undefined,
  },
  "backgroundImage": "url(img.png)",
}
`;

exports[`retinaImage should allow passing in an extension with a dot 1`] = `
Object {
  "
    @media only screen and (-webkit-min-device-pixel-ratio: 1.3),
    only screen and (min--moz-device-pixel-ratio: 1.3),
    only screen and (-o-min-device-pixel-ratio: 1.3/1),
    only screen and (min-resolution: 125dpi),
    only screen and (min-resolution: 1.3dppx)
  ": Object {
    "backgroundImage": "url(img_2x.jpg)",
    "backgroundSize": undefined,
  },
  "backgroundImage": "url(img.jpg)",
}
`;

exports[`retinaImage should set the background-size if one is passed in 1`] = `
Object {
  "
    @media only screen and (-webkit-min-device-pixel-ratio: 1.3),
    only screen and (min--moz-device-pixel-ratio: 1.3),
    only screen and (-o-min-device-pixel-ratio: 1.3/1),
    only screen and (min-resolution: 125dpi),
    only screen and (min-resolution: 1.3dppx)
  ": Object {
    "backgroundImage": "url(img_2x.png)",
    "backgroundSize": "cover",
  },
  "backgroundImage": "url(img.png)",
}
`;

exports[`retinaImage should set the extension if one is passed in 1`] = `
Object {
  "
    @media only screen and (-webkit-min-device-pixel-ratio: 1.3),
    only screen and (min--moz-device-pixel-ratio: 1.3),
    only screen and (-o-min-device-pixel-ratio: 1.3/1),
    only screen and (min-resolution: 125dpi),
    only screen and (min-resolution: 1.3dppx)
  ": Object {
    "backgroundImage": "url(img_2x.jpg)",
    "backgroundSize": undefined,
  },
  "backgroundImage": "url(img.jpg)",
}
`;

exports[`retinaImage should throw an error when a non-string value is passed for backgroundSize 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/retinaImage.js --%c

retinaImage %cexpects a 2nd parameter of type string. However, you passed 100(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#retinaImage%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`retinaImage should throw an error when a non-string value is passed for extension 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/retinaImage.js --%c

retinaImage %cexpects a 3rd parameter of type string. However, you passed true(boolean) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#retinaImage%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`retinaImage should throw an error when a non-string value is passed for fileName 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/retinaImage.js --%c

retinaImage %cexpects a 1st parameter of type string. However, you passed 1(number) instead.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#retinaImage%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`retinaImage should throw an error when no parameters are passed 1`] = `
Array [
  Array [
    "%c ✨ ERROR ✨ ---- mixins/retinaImage.js --%c

retinaImage %cexpects a 1st parameter of type string. However, you did not pass one.
  %c
Please see the documentation at %chttps://polished.js.org/docs/#retinaImage%c for more information.",
    "font-weight: bold; color: black",
    "color: black; font-size: 12px; font-weight: bold",
    "color: black; font-size: 12px",
    "color: gray; line-height: 1.4",
    "color: blue; line-height: 1.4",
    "color: gray; line-height: 1.4",
  ],
]
`;

exports[`retinaImage should use _2x and png as the default suffix and extension, respectively 1`] = `
Object {
  "
    @media only screen and (-webkit-min-device-pixel-ratio: 1.3),
    only screen and (min--moz-device-pixel-ratio: 1.3),
    only screen and (-o-min-device-pixel-ratio: 1.3/1),
    only screen and (min-resolution: 125dpi),
    only screen and (min-resolution: 1.3dppx)
  ": Object {
    "backgroundImage": "url(img_2x.png)",
    "backgroundSize": undefined,
  },
  "backgroundImage": "url(img.png)",
}
`;
